kernel/comps/virtio/src/device/block/device.rs

```diff
@@ -351,28 +394,53 @@ impl DeviceInner {
         };
 
         let inputs = {
-            let mut inputs: Vec<&DmaStreamSlice> = bufs.iter().collect();
-            inputs.insert(0, &req_slice);
+            let mut inputs: Vec<&DmaStreamSlice> = Vec::with_capacity(dma_slices.len() + 1);
+            inputs.push(&req_slice);
+            inputs.extend(dma_slices.iter());
             inputs
         };
 
-        let mut queue = self.queue.lock_irq_disabled();
-        let token = queue
-            .add_dma_buf(inputs.as_slice(), &[&resp_slice])
-            .expect("add queue failed");
-        queue.notify();
-        while !queue.can_pop() {
-            spin_loop();
+        let num_used_descs = inputs.len() + 1;
+        // FIXME: Split the request if it is too big
+        if num_used_descs > Self::QUEUE_SIZE as usize {
+            panic!("The request size surpasses the queue size");
         }
-        queue.pop_used_with_token(token).expect("pop used failed");
+        loop {
+            let mut queue = self.queue.lock_irq_disabled();
+            if num_used_descs > queue.available_desc() {
+                continue;
+            }
+            let token = queue
+                .add_dma_buf(inputs.as_slice(), &[&resp_slice])
+                .expect("add queue failed");
+            if queue.should_notify() {
+                queue.notify();
+            }
 
-        resp_slice.sync().unwrap();
-        let resp: BlockResp = resp_slice.read_val(0).unwrap();
-        self.id_allocator.lock().push(id as u8);
-        match RespStatus::try_from(resp.status).unwrap() {
-            RespStatus::Ok => {}
-            _ => panic!("io error in block device:{:?}", resp.status),
-        };
+            // Records the submitted request
+            let submitted_request = SubmittedRequest::new(id as u16, bio_request, dma_streams);
+            self.submitted_requests
+                .lock()
+                .insert(token, submitted_request);
+            return;
+        }
+    }
+}
+
+#[derive(Debug)]
+struct SubmittedRequest {
+    id: u16,
+    bio_request: BioRequest,
+    bufs: Vec<(DmaStream, usize, usize)>,
```

Comment:
Maybe we should give a better name or type to the tuple (DmaStream, usize, usize).

Reply:
In actuality, it is a `DmsStreamSlice`; however, it needs to maintain ownership of `DmaStream` (this is because `DmaStream` does not retain ownership but is merely a reference).

Reply:
Maybe we should consider implementing an owned type for `DmaStreamSlice` ?

Reply:
I think using a tuple is simple enough to represent the `DmaStream`, and we can use it to construct the `DmaStreamSlice`

Reply:
Okay